engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/stores/ExecutionStore.scala

```diff
@@ -0,0 +1,253 @@
+  protected def withNeedsUpdateTrue: ExecutionStore
+
+  /**
+    * Resets the needsUpdate Boolean to false.
+    */
+  protected def withNeedsUpdateFalse: ExecutionStore
+  
+  /*
+    * Create 2 Tables, one for keys in done status and one for keys in terminal status.
+    * A Table is nothing more than a Map[R, Map[C, V]], see Table trait for more details
+    * In this case, rows are GraphNodes, columns are ExecutionIndexes, and values are JobKeys
+    * This allows for quick lookup of all shards for a node, as well as accessing a specific key with a 
+    * (node, index) pair
+   */
+  lazy val (doneStatus, terminalStatus) = {
+    def toTableEntry(key: JobKey) = (key.node, key.index, key)
+
+    store.foldLeft((Table.empty[GraphNode, ExecutionIndex, JobKey], Table.empty[GraphNode, ExecutionIndex, JobKey]))({
+      case ((done, terminal), (status, keys))  =>
+        lazy val newMapEntries = keys map toTableEntry
+        val newDone = if (status.isDoneOrBypassed) done.addAll(newMapEntries) else done
+        val newTerminal = if (status.isTerminal) terminal.addAll(newMapEntries) else terminal
+
+        newDone -> newTerminal
+    })
+  }
+
+  private def keysWithStatus(status: ExecutionStatus) = store.getOrElse(status, List.empty)
+
+  def isInBypassedConditional(jobKey: JobKey): Boolean = keysWithStatus(Bypassed).exists {
+    case conditional: ConditionalKey if conditional.node.innerGraph.nodes.contains(jobKey.node)
+      && conditional.index.equals(jobKey.index) => true
+    case _ => false
+  }
+
+  /**
+    * We're done when all the keys have a terminal status,
+    * which is equivalent to non of them being in a non-terminal status and faster to verify
+    */
+  def isDone: Boolean = {
+    NonTerminalStatuses.map(keysWithStatus).forall(_.isEmpty)
+  }
+
+  def jobStatus(jobKey: JobKey): Option[ExecutionStatus] = statusStore.get(jobKey)
+
+  def startedJobs: List[BackendJobDescriptorKey] = {
+    store.filterNot({ case (s, _) => s == NotStarted}).values.toList.flatten collect {
+      case k: BackendJobDescriptorKey => k
+    }
+  }
+
+  override def toString: String = store.map { case (j, s) => s"$j -> $s" } mkString System.lineSeparator()
+
+  /**
+    * If needsUpdate is true, goes through NotStarted keys and determines the ones that can be run.
+    * Only computes the first MaxJobsToStartPerTick runnable keys.
+    * In the process, identifies and updates the status of keys that are unreachable.
+    * Returns an ExecutionStoreUpdate which is the list of runnable keys and an updated execution store.
+    * 
+    * If needsUpdate, returns an empty list of runnable keys and this instance of the store.
+    * 
+    * This method can expansive to run for very large workflows if needsUpdate is true.
+    */
+  def update: ExecutionStoreUpdate = if (needsUpdate) {
+    // When looking for runnable keys, keep track of the ones that are unreachable so we can mark them as such
+    var unreachables = Map.empty[JobKey, ExecutionStatus]
```

Comment:
ToL could do a `foldMap` below and skip the var here.

Reply:
I think we'd need some sort of short-circuiting fold if we want to keep the "stream" behavior (see my answer to your stream question) ?